### ISSUE TITLE
V2 and V3 scheduler compatiblity

### ISSUE THREAD

**Issue by rwmpelstilzchen** [oakkitten/anki-delay-siblings#7]
Thanks for making this add-on 🙂

It looks great, but I hesitate to use it because I’m not sure if it is compatible with the V3 scheduler. I don’t want to ruin my decks because of software incompatibility.

Is the add-on compatible with V2 and V3 schedulers? If it is, please indicate so in the add-on page and the readme on GitHub. If it isn’t, please write a warning (and I would appreciate it very much if you make it compatible).

**Comment by oakkitten** [comment]
to be honest i don't know if it's compatible. in all probability it is, as the way it reschedules is quite direct—it changes card's `due` and “flushes” card information. this kind of action doesn't make it into card's history and should be scheduler agnostic. 

said that, i think i didn't test it with the v2+ scheduler. i would like to field test it for a while before saying that it is compatible. this is not something i will be able to do in the nearest future because of reasons, so this issue might stay for a while~

**Comment by rwmpelstilzchen** [comment]
Thanks!

BTW, if the card’s `due` is changed without reflecting it in the card’s history, does it make a discrepancy between the actual due date and the card’s history. I am far from an expert on this, but for example if you sets the due date manually (<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd>) it *is* reflected in the history.

**Comment by oakkitten** [comment]
yeah i recently noticed that `Ctrl`+`Shift`-`D` does make it into history. i didn't go this way as at the time the shortcut didn't exist so i didn't know this was possible. also not entirely sure if this way is a good one as what this plugin might not fit the existing “types” of rescheduling well (Learn, Review, Relearn, Manual, something else maybe). this needs investigating, yea. also, this will probably give us the undo so that's a good thing i guess.

now that you mentioned it, i wonder if the “Check Database” action might be “fixing” the actions of this plugin since as you said there *is* a discrepancy. better check this as well~

**Comment by rwmpelstilzchen** [comment]
An idea: maybe instead of directly setting the due date of cards, which can possibly cause database discrepancies and may clash with the built-in schedulers, we can just bury cards every day. This is the approach taken by [bury future new sibs](https://ankiweb.net/shared/info/930365039) and IMHO it is much safer. Cards can be buried all at once at the beginning of the day or ‘on demand’ while reviewing. This does require some rewriting, but on the long run it can ensure better compatibility and make the add-on future-proof.

**Comment by oakkitten** [comment]
hm, i'm not quite sure how you can regularly get into a situation where you have a mature card and a new card as siblings. that's an interesting approach but i feel like they chose this way only because burying is about the only thing you can do with the new cards i think?

i like actually delaying cards more i guess. for one, delayed siblings stay delayed across devices. see also the other issue. and, you actually see the number of due cards getting smaller and smaller. a major point of this plugin is helping with getting through a backlog of cards and i'm not sure how well burying would work there. the statistics probably would not look good. it *is* safer, but then delay siblings went through enough field testing i think, and if there are problems in the future, well, we can deal with them when we get to them~

**Comment by rwmpelstilzchen** [comment]
> hm, i'm not quite sure how you can regularly get into a situation where you have a mature card and a new card as siblings

If you use that add-on, it takes time for new cards to appear when you have another young card from the same note. It spaces out new cards of the same note: card 2 appears (=is not buried) only when card 1 is mature enough (you decide how many days that is) and is not scheduled too soon; card 3 appears only when both cards 2 and 1 are mature enough and are not scheduled too soon, etc.

> that's an interesting approach but i feel like they chose this way only because burying is about the only thing you can do with the new cards i think?

Possibly. You have to manually run the add-on every day before reviewing.

> a major point of this plugin is helping with getting through a backlog of cards and i'm not sure how well burying would work there.

How is that different from normal use?

**Comment by oakkitten** [comment]
> If you use that add-on, it takes time for new cards to appear when you have another young card from the same note. It spaces out new cards of the same note: card 2 appears (=is not buried) only when card 1 is mature enough (you decide how many days that is) and is not scheduled too soon; card 3 appears only when both cards 2 and 1 are mature enough and are not scheduled too soon, etc.

why would you want to have a situation where a new card and a mature card share the same note? i'm not sure why this much spacing would be desirable. anyway, from the description i gather that this situation is problematic and one they want to avoid, well, at least that's the scenario “this add-on solves the problem with” no?

> How is that different from normal use?

well a while ago i got into a situation where i didn't do reviews for a while and had around 10k cards due. a big portion of these were siblings. if during daily use you may have some 5-6 siblings out of 100 that need to be delayed, during that time that number was in thousands. it's a bit hard to imagine how the statistics would look with burying, when some 3k cards are getting buried at the start of each day. maybe it would work okay but maybe it will be also screwing the statistics idk. with actual rescheduling, well, you get that feeling of getting faster and faster through reviews, as reviewing one card may put more than just one from the overall queue. this gets your spirit up and that's important imo.

btw they do direct sql queries in that plugin and this is something i absolutely want to avoid. dangerous stuff

**Comment by rwmpelstilzchen** [comment]
> why would you want to have a situation where a new card and a mature card share the same note? i'm not sure why this much spacing would be desirable.

Any time you want to have a good grasp of card N before learning card N+1. For example, say you learn a language and use sentence cards. First you want to understand the written version, because — Japanese and Chinese aside — the written form is easier to work with. Only once you can easily comprehend the written sentence you want to proceed to learn to comprehend it by ear. If these two cards enter the cards you review in a single day span (as they will normally do in Anki) you will have to deal with everything in about the same time: not only it is harder, but the feedback you give Anki is problematic because card 1 helps the short-term memory of card 2 and via versa.

> btw they do direct sql queries in that plugin and this is something i absolutely want to avoid. dangerous stuff

Yeah… 😨

**Comment by oakkitten** [comment]
Sorry for a wee delay 😅

I made some automated tests and they pass for both v2 and v3 scheduler. I must say the tests are not ideal; while they do work in a very much actual Anki, it's just so hard to think of behavior that needs testing, apart from the actual “it works” / “it blows up”.

I also looked at what `Ctrl-Shift-D` does and I think the primary purpose of its altering the log is that it provides undo. I think it would be unwise for us to follow this route because:

* There's no “Reschedued by addon” category
* Having undo would be confusing for the user. `Ctrl-Z` normally undoes the last review. The user would be pressing that and wondering why it doesn't show the previous card right away.
* The delay the script introduces is not that big, anyway.

I'll be closing this as the primary issue has been resolved. I won't be releasing a new version on AnkiWeb for this, as most code changes are not such that they change behavior. If you found any more problems, don't hesitate to open a new issue!

P.S. I'm *very* not sold on separating learning stuff in writing and only then by ear. This knowledge complements each other and should be learned together I think.
